### PR TITLE
[DTensor RNG][BC Breaking] Change DTensor Philox seed and offset from int to tensor

### PR DESCRIPTION
**Pull request by yiming0416**
When tracing a DTensor random op, it involves tracing the RNG state in `OffsetBasedRNGTracker`. However the `.item()` call in the `_PhiloxState` is not fake tensor friendly. 

In this PR, we remove the `.item()` call in `_PhiloxState` so that the getter and setter of `seed` and `offset` directly interact with Tensors and it is more tracing friendly.

`_LocalPhiloxState` in `_LocalOffsetBasedRNGTracker` has a similar pattern that we could potentially get rid of the `.item()` call. Right now we don't trace local tensor, so PR keeps it unchanged but makes sure it's compatible with the new `_PhiloxState`

### CODE REVIEW

**Review by dolpm — APPROVED** [review]

File: torch/distributed/_local_tensor/__init__.py [hunk 1]
```diff
@@ -795,8 +795,8 @@ def _distribute_region(self, spec, generator=None):
                 from torch.distributed.tensor._random import _PhiloxState
 
                 any_rank_philox = _PhiloxState(any_rank_device_state)
-                state.seed = any_rank_philox.seed
-                state.offset = any_rank_philox.offset
+                state.seed = int(any_rank_philox.seed.item())
```
Comment (wconstab) [inline torch/distributed/_local_tensor/__init__.py:798]: you're adding a new .item() call in the per op dispatch hot path.  

is this only ok because you aren't tracing localtensor?  does it have perf implications to localtensor mode though?

File: torch/distributed/tensor/_random.py [hunk 2]
```diff
@@ -125,26 +125,20 @@ def state(self):
         return self._state
 
     @property
-    def offset(self) -> int:
-        return int(self._state[8:].view(dtype=torch.int64).item())
+    def offset(self) -> torch.Tensor:
```
Comment (wconstab) [inline torch/distributed/tensor/_random.py:128]: this is technically a BC breaking change, though it is on an underscore API, it is something external users depend on (I would cc @wanchaol and @leonardo0lyj @jc-bytedance @pengyanghua to see if they mind at least).

File: torch/distributed/tensor/_random.py [hunk 3]
```diff
@@ -125,26 +125,20 @@ def state(self):
...
+    def seed(self) -> torch.Tensor:
+        return self._state[:8].view(dtype=torch.int64)
 
     @seed.setter
-    def seed(self, seed: int) -> None:
-        seed_tensor = torch.tensor([seed], dtype=torch.uint64, device="cpu").view(
-            torch.uint8
-        )
-        self._state[:8] = seed_tensor
+    def seed(self, seed: torch.Tensor) -> None:
```
Comment (wconstab) [inline torch/distributed/tensor/_random.py:141]: the new version of the API probably should be more strict: previosuly, it was always going to be a tensor with 1 int64 value in it viewed as uint8.  now, it is not gauranteed to be a 1elt tensor anymore

File: torch/distributed/tensor/_random.py [hunk 4]
```diff
@@ -125,26 +125,20 @@ def state(self):
...
-        offset_tensor = torch.tensor([offset], dtype=torch.uint64, device="cpu").view(
-            torch.uint8
-        )
-        self._state[8:] = offset_tensor
+    def offset(self, offset: torch.Tensor) -> None:
+        self._state[8:] = offset.view(torch.uint8)
 
     @property
-    def seed(self) -> int:
-        return int(self._state[:8].view(dtype=torch.uint64).item())
```
Comment (wconstab) [inline torch/distributed/tensor/_random.py:140]: i'm happy to get rid of this .item() call

**Review by wconstab — COMMENTED** [review]
i think this is probably a good change. it is BC breaking though, so we need to at least document it and ensure our release notes flag it. We should also wait for some others to take a look and give feedback.

File: torch/distributed/_local_tensor/__init__.py (hunk 1 above)
Comment (dolpm) [reply torch/distributed/_local_tensor/__init__.py:798]: local tensor should be fine... wouldn't worry about the overhead here.

Comment (yiming0416) [reply torch/distributed/_local_tensor/__init__.py:798]: Currently we are not tracing `LocalTensor`. I did see  `_LocalPhiloxState` in `_LocalOffsetBasedRNGTracker` which has a similar pattern that we could potentially get rid of `.item()` call there. But it has a larger surface so I didn't do that.

File: torch/distributed/tensor/_random.py (hunk 3 above)
Comment (yiming0416) [reply torch/distributed/tensor/_random.py:141]: added assert to make sure `seed` is a 1-elt tensor

**Comment by yiming0416** [comment]
> can you also update your PR desc with a full description of why/what you changed?

@wconstab updated the PR desc.

**Review by wconstab — APPROVED** [review]
